### ISSUE TITLE
Data race with loaders

### ISSUE THREAD

**Issue by hiendaovinh** [stephenafamo/bob#55]
For now, the random source for sides is defined as
```go
var randsrc = rand.New(rand.NewSource(int64(new(maphash.Hash).Sum64())))
```
When the preloaders are called concurrently, there will be race condition error with concurrent writes at randsrc.
```
WARNING: DATA RACE
Write at 0x00c0003fb500 by goroutine 536:
  math/rand.(*rngSource).Uint64()
      /usr/local/go/src/math/rand/rng.go:239 +0x44
  math/rand.(*rngSource).Int63()
      /usr/local/go/src/math/rand/rng.go:234 +0x1f1
  math/rand.(*Rand).Int63()
      /usr/local/go/src/math/rand/rand.go:84 +0x81
  math/rand.(*Rand).Int63n()
      /usr/local/go/src/math/rand/rand.go:116 +0x4c
  github.com/stephenafamo/bob/dialect/psql.Preload[...].func1()
  
Previous write at 0x00c0003fb500 by goroutine 535:
  math/rand.(*rngSource).Uint64()
      /usr/local/go/src/math/rand/rng.go:239 +0x44
  math/rand.(*rngSource).Int63()
      /usr/local/go/src/math/rand/rng.go:234 +0x1f1
  math/rand.(*Rand).Int63()
      /usr/local/go/src/math/rand/rand.go:84 +0x81
  math/rand.(*Rand).Int63n()
      /usr/local/go/src/math/rand/rand.go:116 +0x4c
  github.com/stephenafamo/bob/dialect/psql.Preload[...].func1()
```

**Comment by hiendaovinh** [comment]
I think we should create independent random sources instead of a singleton source.

**Comment by stephenafamo** [comment]
I missed the note in the package that it should be used only by a single goroutine. I'll change it.

**Comment by stephenafamo** [comment]
Should be fixed in 5c1382d

Try it and let me know.

**Comment by hiendaovinh** [comment]
The hashmap itself is already thread-safe, adding mutex lock could result in lower performance. I think we can just go with it directly instead of using it as the rng seed.
```go
package main

import (
	"fmt"
	"hash/maphash"
	"sync"
)

func rand(n int64) int64 {
	outUint64 := new(maphash.Hash).Sum64()
	out := int64(outUint64)

	if out < 0 {
		return -out % n
	}
	return out % n
}

func main() {
	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			fmt.Println(rand(10000))
		}()
	}

	wg.Wait()
}

```

**Comment by stephenafamo** [comment]
Are we sure that creating a new hashmap is faster than using a mutex? I'll run some benchmarks to see

**Comment by stephenafamo** [comment]
It is faster

```
goos: darwin
goarch: arm64
pkg: github.com/stephenafamo/bob/internal
BenchmarkRandIntMutex
BenchmarkRandIntMutex       	84859948	        13.84 ns/op	       0 B/op	       0 allocs/op
BenchmarkRandIntMutex-2     	82524325	        14.03 ns/op	       0 B/op	       0 allocs/op
BenchmarkRandIntMutex-4     	82061802	        13.99 ns/op	       0 B/op	       0 allocs/op
BenchmarkRandIntHashmap
BenchmarkRandIntHashmap     	129266263	         9.308 ns/op	       0 B/op	       0 allocs/op
BenchmarkRandIntHashmap-2   	129106176	         9.384 ns/op	       0 B/op	       0 allocs/op
BenchmarkRandIntHashmap-4   	129502719	         9.254 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/stephenafamo/bob/internal	11.103s
```

**Comment by stephenafamo** [comment]
Try 447161b

**Comment by hiendaovinh** [comment]
Can you share the benchmark? Thank you.

**Comment by stephenafamo** [comment]
> Can you share the benchmark? Thank you.

Nothing fancy

```go
func BenchmarkRandIntMutex(b *testing.B) {
        for n := 0; n < b.N; n++ {
                RandInt()
        }
}

func BenchmarkRandIntHashmap(b *testing.B) {
        for n := 0; n < b.N; n++ {
                RandInt2()
        }
}
```


```sh
go test -v -run=XXX -cpu 1,2,4 -benchmem -bench=.
```

**Comment by hiendaovinh** [comment]
This is my benchmark
```
goos: linux
goarch: amd64
pkg: [redacted]
cpu: Intel(R) Core(TM) i5-10400 CPU @ 2.90GHz
BenchmarkRandIntMutex
BenchmarkRandIntMutex           45718747                25.14 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntMutex-2         30252022                38.38 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntMutex-4         24955198                49.03 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntHashmap
BenchmarkRandIntHashmap         92770968                12.93 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntHashmap-2       176704159                6.196 ns/op           0 B/op          0 allocs/op
BenchmarkRandIntHashmap-4       383920491                3.148 ns/op           0 B/op          0 allocs/op
PASS
ok      [redacted]    8.205s
```

```go
import (
	"hash/maphash"
	"math/rand"
	"sync"
	"testing"
)

var (
	randsrc = rand.New(rand.NewSource(int64(new(maphash.Hash).Sum64()))) //nolint:gosec
	randmu  sync.Mutex
)

func Rand1() int64 {
	randmu.Lock()
	defer randmu.Unlock()
	return randsrc.Int63n(10000)
}

func Rand2() int64 {
	outUint64 := new(maphash.Hash).Sum64()
	out := int64(outUint64)

	if out < 0 {
		return -out % 10000
	}
	return out % 10000
}

func BenchmarkRandIntMutex(b *testing.B) {
	b.RunParallel(func(bp *testing.PB) {
		for bp.Next() {
			Rand1()
		}
	})
}

func BenchmarkRandIntHashmap(b *testing.B) {
	b.RunParallel(func(bp *testing.PB) {
		for bp.Next() {
			Rand2()
		}
	})
}
```

**Comment by hiendaovinh** [comment]
For reference, this is your benchmark on my machine
```
goos: linux
goarch: amd64
pkg: [redacted]
cpu: Intel(R) Core(TM) i5-10400 CPU @ 2.90GHz
BenchmarkRandIntMutex
BenchmarkRandIntMutex           45968859                24.80 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntMutex-2         43513534                25.09 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntMutex-4         44538752                24.79 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntHashmap
BenchmarkRandIntHashmap         81288492                13.27 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntHashmap-2       83931314                13.05 ns/op            0 B/op          0 allocs/op
BenchmarkRandIntHashmap-4       85700340                12.97 ns/op            0 B/op          0 allocs/op
PASS
ok      [redacted]    6.779s
```

**Comment by hiendaovinh** [comment]
Yeah, look like RandIntHashmap outperforms RandIntMutex.

**Comment by hiendaovinh** [comment]
Even without mutex lock, bare Hashmap still yield better performance. I think it's safe to close the issue.

**Comment by stephenafamo** [comment]
Released in `v0.21.0`
